Civi/Api4/Action/Contact/ContactSaveTrait.php

```diff
@@ -56,14 +56,21 @@ protected function write(array $items) {
   protected function saveLocations(array $params, $contact) {
     foreach (['Address', 'Email', 'Phone', 'IM'] as $entity) {
       foreach (['primary', 'billing'] as $type) {
-        $prefix = strtolower($entity) . '_' . $type . '.';
-        $item = \CRM_Utils_Array::filterByPrefix($params, $prefix);
-        // Not allowed to update by id or alter primary or billing flags
-        unset($item['id'], $item['is_primary'], $item['is_billing']);
+        $joinName = strtolower($entity) . '_' . $type;
+        $joinPrefix = $joinName . '.';
+        $item = \CRM_Utils_Array::filterByPrefix($params, $joinPrefix);
+        // For updating by id, setting e.g. 'primary_email' is equivalent to 'primary_email.id'
+        if (array_key_exists($joinName, $params)) {
+          $item['id'] ??= $params[$joinName];
+        }
+        // Not allowed to alter primary or billing flags
+        unset($item['is_primary'], $item['is_billing']);
         if ($item) {
           $labelField = CoreUtil::getInfoItem($entity, 'label_field');
-          // If NULL was given for the main field (e.g. `email`) then delete the record
-          if ($labelField && array_key_exists($labelField, $item) && is_null($item[$labelField])) {
+          $labelParamExists = $labelField && array_key_exists($labelField, $item);
+          $idParamExists = array_key_exists('id', $item);
```

Comment:
@colemanw could this be `!empty()` - we are seeing breakage in an api call we have that looks like

`Contact::save(FALSE)->addRecord(['address_primary.id' => NULL, 'contact_id' => 123, 'address_primary.street_address' => 'Seasame Street'])->execute();`

I can look at why we are setting `address_primary.id` - probably because on some rows it is not NULL & on others it is but this doesn't seem like an intentional behaviour change in 6.6 or one that would not be easily changed back?